### PR TITLE
fix: await Redis increment pipeline before budget sync reads

### LINKED ISSUE #20886: [Bug]: Budget Limiter Memory Cache Sync Can Read Stale Redis Data
### Check for existing issues

- [x] I have searched the existing issues and checked that my issue is not a duplicate.

### What happened?

The `RouterBudgetLimiting._push_in_memory_increments_to_redis()` method uses `asyncio.create_task()` to schedule Redis increment pipelines without awaiting their completion. The `_sync_in_memory_spend_with_redis()` method then immediately fetches from Redis after awaiting `_push_in_memory_increments_to_redis()`, but the pipeline may not have executed yet, causing memory cache synchronization to use stale data. This leads to budget tracking inconsistencies and potential budget overruns in production proxy deployments with multiple concurrent requests.

**Severity:** 9/10 - Regularly visible malfunction in multi-instance or high-concurrency environments. Budget limits can be exceeded silently due to temporary memory cache staleness between periodic sync operations.

## Root Cause & Impact Flow

- **File(s):**
  - `litellm/router_strategy/budget_limiter.py` (lines 498-502, 530)
  
- **Function/Method:**
  - `_push_in_memory_increments_to_redis()` at lines 481-509
  - Line 498-502 creates unawaited task
  
- **Trigger path:**
  ```
  Multiple concurrent API requests increment budget counters
    → RouterBudgetLimiting.async_log_success_event()
      → _increment_spend_in_current_window() 
        → dual_cache.in_memory_cache.async_increment() [adds to memory]
        → redis_increment_operation_queue.append() [queues for Redis]
          
  Periodic sync (every 1 second):
    → periodic_sync_in_memory_spend_with_redis()
      → await _sync_in_memory_spend_with_redis()
        → await _push_in_memory_increments_to_redis()
          → asyncio.create_task(async_increment_pipeline(...)) [NOT AWAITED!]
            Task scheduled but returns immediately
          → Queue cleared
        → Fetch from Redis: async_batch_get_cache() 
          **BUG:** Pipeline hasn't executed yet, fetching stale data!
        → Update memory with stale value
        
  Subsequent budget check:
    → dual_cache.async_get_cache() reads MEMORY first
      → Gets stale value (lost increments!)
      → Budget check passes incorrectly
  ```

- **What goes wrong:** 
  - `asyncio.create_task()` at line 498 schedules Redis pipeline but returns immediately
  - Method completes, but pipeline still pending in event loop
  - `_sync_in_memory_spend_with_redis()` fetches from Redis expecting recent data
  - Gets stale value because pipeline hasn't executed yet
  - Memory cache is overwritten with this stale value

- **Why it's severe in practice:**
  
  1. **Multi-instance deployments:** Proxy runs with multiple workers, each maintaining separate memory caches synced via Redis
  
  2. **Budget enforcement is critical safety mechanism**:
     - `model_max_budget_limiter.py` tracks spend via `_increment_spend_for_key()` 
     - Budget limits prevent revenue loss from exceeded API quotas
     - If budget tracking loses data, users can exceed allocated budgets
   
  3. **Silent budget overrun:** Memory staleness allows requests based on outdated values
   
  4. **Worst-case scenario (demonstrated below):**
     - Memory shows $160 actual spend
     - Redis pipeline hasn't completed yet
     - Memory synced to stale Redis value ($100)
     - Lost $60 (37.5% data loss!)
     - User can overspend by $60 before next sync

### Steps to Reproduce

Test Script: Budget Sync Staleness Demonstration

```python
import asyncio
import sys

from litellm.router_strategy.budget_limiter import RouterBudgetLimiting
from litellm.caching.caching import DualCache


async def test():
    """Demonstrate budget limiter memory sync bug"""

    # Track values separately for Redis and Memory to show the bug
    redis_values = {"provider_spend:openai:1d": 100.0}

    class MockRedis:
        async def async_get_cache(self, key, **kwargs):
            return redis_values.get(key)

        async def async_batch_get_cache(self, key_list, **kwargs):
            return {k: redis_values.get(k) for k in key_list}

        async def async_set_cache(self, key, value, **kwargs):
            redis_values[key] = float(value)

        async def async_increment_pipeline(self, increment_list, **kwargs):
            # Simulate 300ms network delay
            await asyncio.sleep(0.3)
            for op in increment_list:
                current = redis_values.get(op["key"], 0)
                new_value = current + op["increment_value"]
                redis_values[op["key"]] = new_value

    class MockMemory:
        def __init__(self):
            self.memory_values = {"provider_spend:openai:1d": 100.0}

        async def async_get_cache(self, key, **kwargs):
            return self.memory_values.get(key)

        async def async_set_cache(self, key, value, **kwargs):
            print(f"    → Memory cache updated ({key}) to {value}")
            self.memory_values[key] = float(value)

        async def async_increment(self, key, value, **kwargs):
            current = self.memory_values.get(key, 0.0) or 0.0
            new_value = float(current) + float(value)
            await self.async_set_cache(key=key, value=new_value, **kwargs)
            return new_value

    mock_redis = MockRedis()
    mock_memory = MockMemory()
    dual_cache = DualCache(in_memory_cache=mock_memory, redis_cache=mock_redis)

    router_budget = RouterBudgetLimiting(
        dual_cache=dual_cache,
        provider_budget_config={"openai": {"budget_limit": 500.0, "time_period": "1d"}},
    )

    # Simulate two rapid requests on the real RouterBudgetLimiting path:
    # this updates memory AND appends Redis increment ops to the queue.
    await router_budget._increment_spend_in_current_window(
        spend_key="provider_spend:openai:1d",
        response_cost=25.0,
        ttl=86400,
    )
    await router_budget._increment_spend_in_current_window(
        spend_key="provider_spend:openai:1d",
        response_cost=35.0,
        ttl=86400,
    )

    print(
        f"Memory after increments: ${mock_memory.memory_values['provider_spend:openai:1d']}"
    )  # $160
    print(
        f"Redis still shows: ${redis_values.get('provider_spend:openai:1d')}"
    )  # $100 (pending sync)

    # Trigger periodic sync - THIS IS WHERE BUG MANIFESTS!
    await router_budget._sync_in_memory_spend_with_redis()

    # Give scheduled Redis pipeline task time to finish, so we can compare eventual Redis value.
    await asyncio.sleep(0.35)

    final_redis = redis_values.get("provider_spend:openai:1d")
    final_memory = mock_memory.memory_values.get("provider_spend:openai:1d")

    print(f"\nAfter sync:")
    print(f"  Redis (eventual): ${final_redis}")
    print(f"  Memory (stale): ${final_memory}")

    if final_memory < final_redis:
        loss = final_redis - final_memory
        loss_pct = (loss / final_redis) * 100
        print(f"\n❌ BUG CONFIRMED!")
        print(f"   Lost ${loss} ({loss_pct:.1f}% data loss)")
        return False
    else:
        print("\n✓ No bug")
        return True


result = asyncio.run(test())
sys.exit(0 if result else 1)

```

### Observed Result (Before Fix)

```bash
    → Memory cache updated (provider_spend:openai:1d) to 125.0
    → Memory cache updated (provider_spend:openai:1d) to 160.0
Memory after increments: $160.0
Redis still shows: $100.0
    → Memory cache updated (provider_spend:openai:1d) to 100.0
    → Memory cache updated (provider_spend:openai:1d) to 100.0
    → Memory cache updated (provider_budget_start_time:openai) to 1770750694.843928

After sync:
  Redis (eventual): $160.0
  Memory (stale): $100.0

❌ BUG CONFIRMED!
   Lost $60.0 (37.5% data loss)

```

### Expected Result

- After periodic sync, memory cache should match Redis exactly ($160)
- No increments should be lost during the sync process
- Budget tracking should be accurate at all times

### Relevant log output

```shell

```

### What part of LiteLLM is this about?

Proxy

### What LiteLLM version are you on ?

v1.81.8

### Twitter / LinkedIn details

_No response_

### PR DESCRIPTION
**Pull request by emerzon**
## Relevant issues

Fixes #20886

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [ ] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## CI (LiteLLM team)

- [ ] **Branch creation CI run**
       Link:

- [ ] **CI run for the last commit**
       Link:

- [ ] **Merge / cherry-pick CI run**
       Links:

## Type

🐛 Bug Fix
✅ Test

## Changes

- Fixes a race in `RouterBudgetLimiting._push_in_memory_increments_to_redis()` where Redis increment pipeline writes were scheduled via `asyncio.create_task(...)` and not awaited.
- Uses an atomic queue snapshot + swap before awaiting Redis flush:
  - snapshot current queue to `increment_operations_to_flush`
  - reset live queue immediately so new increments are queued for next cycle
  - await `async_increment_pipeline(...)` for the snapshot before continuing
- Adds defensive behavior: if Redis flush fails, re-queue the flushed increments to avoid dropping spend updates.
- Adds regression tests in `tests/test_litellm/router_strategy/test_budget_limiter.py`:
  - verifies `_sync_in_memory_spend_with_redis()` waits for Redis pipeline completion before Redis reads
  - verifies increments are re-queued when Redis pipeline write fails

## Validation

Ran locally:

- `pytest -q tests/test_litellm/router_strategy/test_budget_limiter.py`
- `pytest -q tests/test_litellm/router_strategy`

Both passed.

### CODE REVIEW

File: litellm/router_strategy/budget_limiter.py [hunk 1]
```diff
@@ -486,24 +486,31 @@ async def _push_in_memory_increments_to_redis(self):
...
         """
+        increment_operations_to_flush: List[RedisPipelineIncrementOperation] = []
         try:
             if not self.dual_cache.redis_cache:
                 return  # Redis is not initialized
 
+            # Snapshot pending increments and clear queue before await, so new writes are queued
+            # for the next sync cycle while this batch is flushed to Redis.
+            increment_operations_to_flush = self.redis_increment_operation_queue
+            self.redis_increment_operation_queue = []
```
[the remark this answers is not in the document]
Comment (emerzon) [reply litellm/router_strategy/budget_limiter.py]: Added queue lock protection. Queue snapshot/swap is now done under `self._redis_increment_queue_lock` in `_push_in_memory_increments_to_redis()`, and queue append in `_increment_spend_in_current_window()` is also lock-protected to avoid interleaving issues.

File: tests/test_litellm/router_strategy/test_budget_limiter.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+class _MockRedisCache:
+    def __init__(self, initial_values):
+        self.values = initial_values
+        self.events = []
+
+    async def async_increment_pipeline(self, increment_list, **kwargs):
+        self.events.append("increment_pipeline:start")
+        await asyncio.sleep(0.05)
+        for op in increment_list:
```
[the remark this answers is not in the document]
Comment (emerzon) [reply tests/test_litellm/router_strategy/test_budget_limiter.py:33]: Good call. Replaced timing-based sleep with `asyncio.Event` gates. The test now deterministically blocks pipeline completion, asserts `batch_get` does not occur early, then releases the event and verifies final ordering.

File: litellm/router_strategy/budget_limiter.py [hunk 3]
```diff
@@ -486,24 +486,31 @@ async def _push_in_memory_increments_to_redis(self):
...
-            self.redis_increment_operation_queue = []
-
         except Exception as e:
+            if len(increment_operations_to_flush) > 0:
+                # Retry these increments on a future sync cycle.
+                self.redis_increment_operation_queue = (
+                    increment_operations_to_flush + self.redis_increment_operation_queue
+                )
             verbose_router_logger.error(
                 f"Error syncing in-memory cache with Redis: {str(e)}"
```
[the remark this answers is not in the document]
Comment (emerzon) [reply litellm/router_strategy/budget_limiter.py]: Updated exception handling here. The log message now reflects the operation (`Error pushing queued Redis increment operations to Redis`) and uses `verbose_router_logger.exception(..., exc_info=True)` so the traceback is preserved.

File: litellm/router_strategy/budget_limiter.py [hunk 4]
```diff
@@ -486,24 +486,31 @@ async def _push_in_memory_increments_to_redis(self):
...
+                    increment_list=increment_operations_to_flush,
                 )
 
-            self.redis_increment_operation_queue = []
-
         except Exception as e:
+            if len(increment_operations_to_flush) > 0:
+                # Retry these increments on a future sync cycle.
+                self.redis_increment_operation_queue = (
+                    increment_operations_to_flush + self.redis_increment_operation_queue
```
[the remark this answers is not in the document]
Comment (emerzon) [reply litellm/router_strategy/budget_limiter.py]: Updated this path as well: re-queuing on Redis flush failure now happens under the same queue lock, and I added a regression test that appends a new increment while a failing flush is in-flight to verify both old and new increments are retained in order.

File: litellm/router_strategy/budget_limiter.py [hunk 5]
```diff
@@ -348,7 +349,17 @@ async def _increment_spend_in_current_window(
...
+            self.redis_increment_operation_queue.append(increment_op)
+
+    def _get_redis_increment_queue_lock(self) -> asyncio.Lock:
+        queue_lock: Optional[asyncio.Lock] = getattr(
+            self, "_redis_increment_queue_lock", None
+        )
+        if queue_lock is None:
+            queue_lock = asyncio.Lock()
+            self._redis_increment_queue_lock = queue_lock
+        return queue_lock
```
[the remark this answers is not in the document]
Comment (emerzon) [reply litellm/router_strategy/budget_limiter.py]: Good point. I simplified `_get_redis_increment_queue_lock()` to directly return `self._redis_increment_queue_lock` and removed lazy creation. The lock remains initialized in `__init__()` and tests that instantiate via `__new__` set the lock explicitly.

**Comment by emerzon** [comment]
@greptile

**Comment by emerzon** [comment]
Superseeded by #32618
